dist/tools/ethos/ethos.c

```diff
@@ -0,0 +1,378 @@
+                TRACE("frame start");
+                serial->state = IN_FRAME;
+                serial->frametype = 0;
+                serial->framebytes = 0;
+            }
+            break;
+        case IN_FRAME:
+            if (c == LINE_ESC_CHAR) {
+                TRACE("i esc");
+                serial->state = IN_ESCAPE;
+            }
+            else if (c == LINE_FRAME_DELIMITER) {
+                if (!serial->framebytes) {
+                    TRACE("i del unexp");
+                } else {
+                    TRACE("i del");
+                    serial->state = WAIT_FRAMESTART;
+                    return serial->framebytes;
+                }
+            }
+            else {
+                _handle_char(serial, c);
+            }
+            break;
+        case IN_ESCAPE:
+            if (c == (LINE_FRAME_DELIMITER ^ 0x20)) {
+                _handle_char(serial, LINE_FRAME_DELIMITER);
+            }
+            else if (c == (LINE_ESC_CHAR ^ 0x20)) {
+                _handle_char(serial, LINE_ESC_CHAR);
+            }
+            else if (c == (LINE_FRAME_TYPE_TEXT ^ 0x20)) {
+                TRACE("esc txt");
+                serial->frametype = LINE_FRAME_TYPE_TEXT;
+            }
+            else if (c == (LINE_FRAME_TYPE_HELLO ^ 0x20)) {
+                TRACE("esc hello");
+                serial->frametype = LINE_FRAME_TYPE_HELLO;
+            }
+            else if (c == (LINE_FRAME_TYPE_HELLO_REPLY ^ 0x20)) {
+                TRACE("esc hello_reply");
+                serial->frametype = LINE_FRAME_TYPE_HELLO_REPLY;
+            }
+            else if (c == LINE_FRAME_DELIMITER) {
+                TRACE("esc -del");
+            }
+            serial->state = IN_FRAME;
+            break;
+    }
+
+    return 0;
+}
+
+static void _write_escaped(int fd, char* buf, ssize_t n)
+{
+    uint8_t out[2];
+    size_t escaped = 0;
+
+    while(n--) {
+        char c = *buf++;
+        switch(c) {
+            case LINE_FRAME_DELIMITER:
+                out[0] = LINE_ESC_CHAR;
+                out[1] = (LINE_FRAME_DELIMITER ^ 0x20);
+                write(fd, out, 2);
+                escaped++;
+                break;
+            case LINE_ESC_CHAR:
+                out[0] = LINE_ESC_CHAR;
+                out[1] = (LINE_ESC_CHAR ^ 0x20);
+                write(fd, out, 2);
+                escaped++;
+                break;
+            default:
+                write(fd, &c, 1);
+        }
+    }
+}
+
+static void _send_hello(int serial_fd, serial_t *serial, unsigned type)
+{
+    char delim = LINE_FRAME_DELIMITER;
+    char head[] = { LINE_FRAME_DELIMITER, LINE_ESC_CHAR, (type ^ 0x20) };
+    write(serial_fd, head, sizeof(head));
+    write(serial_fd, serial->local_l2_addr, 6);
+    write(serial_fd, &delim, 1);
+}
+
+static void _clear_neighbor_cache(const char *ifname)
+{
+    char tmp[20 + IFNAMSIZ];
+    snprintf(tmp, sizeof(tmp), "ip neigh flush dev %s", ifname);
```

Comment:
the flush flushes positive cache entries.
You can see the effect of the flush call when you
1. startup ethos
2. ping from linux
3. while linux is pinging, type "reboot" at ethos shell

The linux ping should continue seamlessly with that flush call enabled, if you comment it out, the linux ping stops after RIOT's reboot as RIOT's MAC address has changed.